examples/android/OGVideoProcDroid/src/ogles_gpgpu/examples/ogvideoprocdroid/CameraProcThread.java

```diff
@@ -0,0 +1,404 @@
+        releaseGL();
+
+        Log.i(CamActivity.TAG, "stopped thread CameraProcThread");
+    }
+
+    private void mainLoop() throws Exception {
+        while (running) {
+            boolean frameAvailable = false;
+            synchronized (lock) {
+                if (isSurfaceChanged) {
+                    // Rare event, process immediately
+                    ogWrapper.setRenderDisp(newWidth, newHeight, OGJNIWrapper.ORIENTATION_FLIPPED);
+                    ogWrapper.prepare(camPreviewFrameSize.width, camPreviewFrameSize.height, false);
+                    isSurfaceChanged = false;
+                }
+                else if (modeChanged) {
+                    // Rare event, process immediately
+                    if (newModeInput) {
+                        ogWrapper.setRenderDispShowMode(OGJNIWrapper.RENDER_DISP_MODE_INPUT);
+                    } else {
+                        ogWrapper.setRenderDispShowMode(OGJNIWrapper.RENDER_DISP_MODE_OUTPUT);
+                    }
+                    modeChanged = false;
+                }
+                else if (frameNotification) {
+                    // Frequent event, do release lock as soon as possible so if one more
+                    // call to onFrameAvailable occurs then there will be
+                    // no need to wait.
+                    frameAvailable = true;
+                    frameNotification = false;
+                }
+                else {
+                    // no event to process, waiting...
+                    lock.wait();
+                }
+            }
+            if (frameAvailable) {
+                processFrame();
+            }
+        }
+    }
+
+    /**
+     * Callback from SurfaceTexture.OnFrameAvailableListener. Is called each time a new
+     * camera frame for the SurfaceTexture is available.
+     */
+    @Override
+    public void onFrameAvailable(SurfaceTexture surfaceTexture) {
+        synchronized (lock) {
+            frameNotification = true;
+            lock.notify();
+        }
+    }
+
+    /**
+      * Process button clicked from main UI.
+      */
+    public void topBtnClicked(boolean modeInput) {
+      synchronized (lock) {
+        modeChanged = true;
+        newModeInput = modeInput;
+        lock.notify();
+      }
+    }
+
+    /**
+      * Process surfaceChanged callback from main UI.
+      */
+    public void surfaceChanged(int width, int height) {
+        synchronized (lock) {
+            newWidth = width;
+            newHeight = height;
+            isSurfaceChanged = true;
+            lock.notify();
+        }
+    }
+
+    /**
+     * Stop the thread.
+     */
+    public void terminate() {
+        synchronized (lock) {
+            running = false;
+            lock.notify();
+        }
+    }
+
+    private void processFrame() {
+        // update camera frame texture
+        GLES20.glActiveTexture(GLES20.GL_TEXTURE1);
+
+        camTexture.updateTexImage();
```

Comment:
As you can see there is no need to call `windowSurface.makeCurrent();` before `updateTexImage`.